### PR TITLE
Multiple stuff

### PR DESCRIPTION
**Pull request by satya164**

### CODE REVIEW

File: src/react/styled.js [hunk 1]
```diff
@@ -93,7 +123,7 @@ type StyledTag<T> = (
 
 type StyledJSXIntrinsics = $ObjMap<
   $JSXIntrinsics,
-  <T>({ props: T }) => StyledTag<T>
+  () => StyledTag<{ children?: React$Node, [key: string]: any }>
```
Comment (satya164) [inline src/react/styled.js]: flow was failing here and due to incomplete dom typings in flow, the same thing would've been the result

File: src/react/styled.js [hunk 2]
```diff
@@ -44,27 +63,38 @@ function styled(tag: React.ComponentType<*> | string) {
...
 
       /* $FlowFixMe */
-      if (typeof tag.className === 'string' && tag !== component) {
+      if (tag.__linaria && tag !== component) {
         // If the underlying tag is a styled component, forward the `as` prop
         // Otherwise the styles from the underlying component will be ignored
         return React.createElement(
           tag,
-          Object.assign(rest, {
+          Object.assign(filteredProps, {
```
Comment (thymikee) [inline src/react/styled.js]: this mutation is on purpose? would be cool to make it explicit

File: src/react/styled.js [hunk 3]
```diff
@@ -44,27 +63,38 @@ function styled(tag: React.ComponentType<*> | string) {
...
+      ? React.forwardRef(render)
+      : // React.forwardRef won't available on older React versions and in Preact
+        // Fallback to a innerRef prop in that case
+        ({ innerRef, ...rest }) => render(rest, innerRef);
 
     Result.displayName = options.name;
-    Result.className = options.class;
-    Result.extends = tag;
+
+    // This properties will be read by the babel plugin for interpolation
```
Comment (thymikee) [inline src/react/styled.js]:
```suggestion
    // These properties will be read by the babel plugin for interpolation
```

**Comment by satya164** [comment]
> what about the footprint of adding is-valid-prop?

what can we do about it though?

**Comment by thymikee** [comment]
nothing, just curious what it is right now :D

**Pull request merged**